File: packages/whisker-firebase/analytics/ios/Sources/WhiskerFirebaseAnalytics/FirebaseAnalyticsModule.swift
import Foundation
import FirebaseAnalytics
import FirebaseCore
import WhiskerModule

@WhiskerModule
public final class FirebaseAnalyticsModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("FirebaseAnalytics")
            Function("logEvent") { (args: [WhiskerValue]) -> WhiskerValue in
                guard args.count == 2, let name = args[0].asString, case .map(let params) = args[1] else {
                    return Self.failure("invalid-argument", "Expected an event name and parameters")
                }
                return Self.configured { Analytics.logEvent(name, parameters: Self.decode(params)) }
            }
            Function("setUserId") { (args: [WhiskerValue]) -> WhiskerValue in
                Self.configured { Analytics.setUserID(args.first?.asString) }
            }
            Function("setUserProperty") { (args: [WhiskerValue]) -> WhiskerValue in
                guard args.count == 2, let name = args[0].asString else {
                    return Self.failure("invalid-argument", "Expected a user property name and value")
                }
                return Self.configured { Analytics.setUserProperty(args[1].asString, forName: name) }
            }
            Function("setAnalyticsCollectionEnabled") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let enabled = args.first?.asBool else {
                    return Self.failure("invalid-argument", "Expected a boolean")
                }
                return Self.configured { Analytics.setAnalyticsCollectionEnabled(enabled) }
            }
            Function("setDefaultEventParameters") { (args: [WhiskerValue]) -> WhiskerValue in
                guard case .map(let params) = args.first else {
                    return Self.failure("invalid-argument", "Expected parameters")
                }
                return Self.configured { Analytics.setDefaultEventParameters(params.isEmpty ? nil : Self.decode(params)) }
            }
            Function("resetAnalyticsData") { (_: [WhiskerValue]) -> WhiskerValue in
                Self.configured { Analytics.resetAnalyticsData() }
            }
            AsyncFunction("getAppInstanceId") { (_: [WhiskerValue], promise: WhiskerPromise) in
                guard FirebaseApp.app() != nil else { promise.resolve(Self.notConfigured()); return }
                promise.resolve(Self.success(Analytics.appInstanceID().map { .string($0) } ?? .null))
            }
            Function("setConsent") { (args: [WhiskerValue]) -> WhiskerValue in
                guard case .map(let fields) = args.first else {
                    return Self.failure("invalid-argument", "Expected consent settings")
                }
                let types: [String: ConsentType] = [
                    "analytics_storage": .analyticsStorage, "ad_storage": .adStorage,
                    "ad_user_data": .adUserData, "ad_personalization": .adPersonalization,
                ]
                var consent: [ConsentType: ConsentStatus] = [:]
                for (key, value) in fields {
                    guard let type = types[key], let granted = value.asBool else { continue }
                    consent[type] = granted ? .granted : .denied
                }
                return Self.configured { Analytics.setConsent(consent) }
            }
            Function("setSessionTimeout") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let millis = args.first?.asInt else {
                    return Self.failure("invalid-argument", "Expected milliseconds")
                }
                return Self.configured { Analytics.setSessionTimeoutInterval(TimeInterval(millis) / 1000) }
            }
        }
    }

    private static func configured(_ body: () -> Void) -> WhiskerValue {
        guard FirebaseApp.app() != nil else { return notConfigured() }
        body()
        return success(.null)
    }

    private static func decode(_ params: [String: WhiskerValue]) -> [String: Any] {
        params.compactMapValues { value -> Any? in
            switch value {
            case .string(let v): return v
            case .int(let v): return NSNumber(value: v)
            case .float(let v): return NSNumber(value: v)
            case .array(let items):
                return items.compactMap { item -> [String: Any]? in
                    guard case .map(let fields) = item else { return nil }
                    return decode(fields)
                }
            default: return nil
            }
        }
    }

    private static func success(_ value: WhiskerValue) -> WhiskerValue { .map(["value": value]) }
    private static func failure(_ code: String, _ message: String) -> WhiskerValue {
        .map(["error": .map(["code": .string(code), "message": .string(message)])])
    }
    private static func notConfigured() -> WhiskerValue {
        failure("app-not-configured", "Call Analytics::instance() before using Analytics")
    }
}

File: packages/whisker-firebase/analytics/android/src/main/kotlin/rs/whisker/modules/firebaseanalytics/FirebaseAnalyticsModule.kt
package rs.whisker.modules.firebaseanalytics

import android.os.Bundle
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import rs.whisker.runtime.Module
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.WhiskerValue

@WhiskerModule
class FirebaseAnalyticsModule : Module() {
    private val analytics get() = FirebaseAnalytics.getInstance(FirebaseApp.getInstance().applicationContext)

    override fun definition() = ModuleDefinition {
        Name("FirebaseAnalytics")
        Function("logEvent") { args ->
            attempt {
                val params = requireNotNull(args.getOrNull(1) as? WhiskerValue.Map).value
                analytics.logEvent(requireNotNull(args.firstOrNull()?.asString()), bundle(params))
            }
        }
        Function("setUserId") { args -> attempt { analytics.setUserId(args.firstOrNull()?.asString()) } }
        Function("setUserProperty") { args ->
            attempt { analytics.setUserProperty(requireNotNull(args.firstOrNull()?.asString()), args.getOrNull(1)?.asString()) }
        }
        Function("setAnalyticsCollectionEnabled") { args ->
            attempt { analytics.setAnalyticsCollectionEnabled(requireNotNull(args.firstOrNull()?.asBool())) }
        }
        Function("setDefaultEventParameters") { args ->
            attempt {
                val params = requireNotNull(args.firstOrNull() as? WhiskerValue.Map).value
                analytics.setDefaultEventParameters(if (params.isEmpty()) null else bundle(params))
            }
        }
        Function("resetAnalyticsData") { _ -> attempt { analytics.resetAnalyticsData() } }
        AsyncFunction("getAppInstanceId") { _, promise ->
            try {
                analytics.appInstanceId.addOnCompleteListener { task ->
                    promise.resolve(
                        if (task.isSuccessful) success(task.result?.let { WhiskerValue.Str(it) } ?: WhiskerValue.Null)
                        else failure(task.exception),
                    )
                }
            } catch (error: Exception) { promise.resolve(failure(error)) }
        }
        Function("setConsent") { args ->
            attempt {
                val fields = requireNotNull(args.firstOrNull() as? WhiskerValue.Map).value
                val types = mapOf(
                    "analytics_storage" to FirebaseAnalytics.ConsentType.ANALYTICS_STORAGE,
                    "ad_storage" to FirebaseAnalytics.ConsentType.AD_STORAGE,
                    "ad_user_data" to FirebaseAnalytics.ConsentType.AD_USER_DATA,
                    "ad_personalization" to FirebaseAnalytics.ConsentType.AD_PERSONALIZATION,
                )
                analytics.setConsent(fields.entries.mapNotNull { (key, value) ->
                    val type = types[key] ?: return@mapNotNull null
                    val granted = value.asBool() ?: return@mapNotNull null
                    type to if (granted) FirebaseAnalytics.ConsentStatus.GRANTED else FirebaseAnalytics.ConsentStatus.DENIED
                }.toMap())
            }
        }
        Function("setSessionTimeout") { args ->
            attempt { analytics.setSessionTimeoutDuration(requireNotNull(args.firstOrNull()?.asInt())) }
        }
    }

    private fun attempt(body: () -> Unit): WhiskerValue =
        try { body(); success(WhiskerValue.Null) } catch (error: Exception) { failure(error) }

    private fun bundle(params: Map<String, WhiskerValue>): Bundle = Bundle().apply {
        for ((key, value) in params) {
            when (value) {
                is WhiskerValue.Str -> putString(key, value.value)
                is WhiskerValue.Int -> putLong(key, value.value)
                is WhiskerValue.Float -> putDouble(key, value.value)
                is WhiskerValue.Array -> putParcelableArray(
                    key,
                    value.value.mapNotNull { (it as? WhiskerValue.Map)?.value?.let(::bundle) }.toTypedArray(),
                )
                else -> {}
            }
        }
    }

    private fun success(value: WhiskerValue) = WhiskerValue.Map(mapOf("value" to value))
    private fun failure(error: Exception?): WhiskerValue {
        val code = when (error) {
            is IllegalArgumentException -> "invalid-argument"
            is IllegalStateException -> "app-not-configured"
            else -> "unknown"
        }
        return WhiskerValue.Map(mapOf("error" to WhiskerValue.Map(mapOf(
            "code" to WhiskerValue.Str(code),
            "message" to WhiskerValue.Str(error?.message ?: "Analytics operation failed"),
        ))))
    }
}
